controlplane/kubeadm/internal/control_plane_test.go

```diff
@@ -252,6 +428,94 @@ func TestHasHealthyMachineStillProvisioning(t *testing.T) {
 	})
 }
 
+func TestMachineInFailureDomainWithMostMachines(t *testing.T) {
+	t.Run("Machines in Failure Domain", func(t *testing.T) {
+		machines := collections.Machines{
+			"machine-3": &clusterv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{Name: "m3"},
+				Spec: clusterv1.MachineSpec{
+					Version:           ptr.To("v1.31.0"),
+					FailureDomain:     ptr.To("three"),
+					InfrastructureRef: corev1.ObjectReference{Kind: "GenericInfrastructureMachine", APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1", Name: "m3"},
+				}},
+		}
+
+		c := &ControlPlane{
+			KCP: &controlplanev1.KubeadmControlPlane{},
+			Cluster: &clusterv1.Cluster{
+				Status: clusterv1.ClusterStatus{
+					FailureDomains: clusterv1.FailureDomains{
+						"three": failureDomain(false),
+					},
+				},
+			},
+			Machines: collections.Machines{
+				"machine-3": machine("machine-3", withFailureDomain("three")),
+			},
+		}
+
+		g := NewWithT(t)
+		_, err := c.MachineInFailureDomainWithMostMachines(ctx, machines)
+		g.Expect(err).NotTo(HaveOccurred())
+	})
+	t.Run("Return error when no controlplane machine found", func(t *testing.T) {
+		machines := collections.Machines{}
+
+		c := &ControlPlane{
+			KCP: &controlplanev1.KubeadmControlPlane{},
+			Cluster: &clusterv1.Cluster{
+				Status: clusterv1.ClusterStatus{
+					FailureDomains: clusterv1.FailureDomains{},
+				},
+			},
+			Machines: collections.Machines{},
+		}
+
+		g := NewWithT(t)
+		_, err := c.MachineInFailureDomainWithMostMachines(ctx, machines)
+		g.Expect(err).To(HaveOccurred())
+	})
+}
+func TestMachineWithDeleteAnnotation(t *testing.T) {
+	t.Run("Machines having delete annotation set", func(t *testing.T) {
+		machines := collections.Machines{
+			"machine-1": &clusterv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{Name: "m1",
+					Annotations: map[string]string{
+						"cluster.x-k8s.io/delete-machine": "",
+					},
+				},
+				Spec: clusterv1.MachineSpec{
+					Version:           ptr.To("v1.31.0"),
+					FailureDomain:     ptr.To("one"),
+					InfrastructureRef: corev1.ObjectReference{Kind: "GenericInfrastructureMachine", APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1", Name: "m1"},
+				}},
+			"machine-2": &clusterv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{Name: "m2",
+					Annotations: map[string]string{
+						"cluster.x-k8s.io/delete-machine": "",
+					},
+				},
+				Spec: clusterv1.MachineSpec{
+					Version:           ptr.To("v1.31.0"),
+					FailureDomain:     ptr.To("two"),
+					InfrastructureRef: corev1.ObjectReference{Kind: "GenericInfrastructureMachine", APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1", Name: "m2"},
+				}},
+		}
+
+		c := ControlPlane{
+			Machines: machines,
+			Cluster: &clusterv1.Cluster{
+				Status: clusterv1.ClusterStatus{},
+			},
+		}
+
+		g := NewWithT(t)
+		annotedMachines := c.MachineWithDeleteAnnotation(machines)
+		g.Expect(annotedMachines).NotTo(BeNil())
```

Comment:
Could we match against `annotatedMachines.Len()`? https://github.com/kubernetes-sigs/cluster-api/blob/main/util/collections/machine_collection.go#L174